### PR TITLE
fix: Remove unnecessary lock in metrics parsing allow-list manifest

### LINKED ISSUE #127992: Components hang forever due to bug in parsing allow metric labels manifest
### What happened?

The components hang forever when initializing with `allow-metric-labels-manifest` is set.

For example, kube-scheduler set the flag as follows
```
root@dev-control-plane:/etc/kubernetes/manifests# cat kube-scheduler.yaml |grep -A8 command
  - command:
    - kube-scheduler
    - --authentication-kubeconfig=/etc/kubernetes/scheduler.conf
    - --authorization-kubeconfig=/etc/kubernetes/scheduler.conf
    - --bind-address=127.0.0.1
    - --kubeconfig=/etc/kubernetes/scheduler.conf
    - --leader-elect=true
    - --allow-metric-labels-manifest=/etc/kubernetes/allow-list.yaml
    image: registry.k8s.io/kube-scheduler:v1.30.0
```
It hangs forever
```
root@dev-control-plane:/etc/kubernetes# kubectl logs -n kube-system kube-scheduler-dev-control-plane -f
I1011 00:10:55.222334       1 serving.go:380] Generated self-signed cert in-memory



^C
```

It should be due to the bug in function [SetLabelAllowListFromManifest](https://github.com/kubernetes/kubernetes/blob/c15581b277e9e42d830898357fdbd92dcf7981c6/staging/src/k8s.io/component-base/metrics/opts.go#L366) which locks the `allowListLock` and tries to double lock in [here](https://github.com/kubernetes/kubernetes/blob/c15581b277e9e42d830898357fdbd92dcf7981c6/staging/src/k8s.io/component-base/metrics/opts.go#L379)

### What did you expect to happen?

The component start and run properly. e.g.
```
root@dev-control-plane:/etc/kubernetes/manifests# kubectl logs -n kube-system kube-scheduler-dev-control-plane -f
I1011 00:20:14.946492       1 serving.go:380] Generated self-signed cert in-memory
I1011 00:20:15.109004       1 server.go:154] "Starting Kubernetes Scheduler" version="v1.30.0"
I1011 00:20:15.109046       1 server.go:156] "Golang settings" GOGC="" GOMAXPROCS="" GOTRACEBACK=""
I1011 00:20:15.113642       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I1011 00:20:15.113671       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I1011 00:20:15.113643       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I1011 00:20:15.113762       1 secure_serving.go:213] Serving securely on 127.0.0.1:10259
I1011 00:20:15.113802       1 shared_informer.go:313] Waiting for caches to sync for RequestHeaderAuthRequestController
I1011 00:20:15.113869       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I1011 00:20:15.113668       1 shared_informer.go:313] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1011 00:20:15.113707       1 shared_informer.go:313] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1011 00:20:15.214278       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I1011 00:20:15.214325       1 leaderelection.go:250] attempting to acquire leader lease kube-system/kube-scheduler...
I1011 00:20:15.214474       1 shared_informer.go:320] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I1011 00:20:15.214495       1 shared_informer.go:320] Caches are synced for RequestHeaderAuthRequestController
I1011 00:20:32.976351       1 leaderelection.go:260] successfully acquired lease kube-system/kube-scheduler```

### How can we reproduce it (as minimally and precisely as possible)?

Mount the correct format allow-list manifest to the pod and set the `--allow-metric-label-manfiests` flag
e.g. static kube-scheduler manfiest
```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    component: kube-scheduler
    tier: control-plane
  name: kube-scheduler
  namespace: kube-system
spec:
  containers:
  - command:
    - kube-scheduler
    - --authentication-kubeconfig=/etc/kubernetes/scheduler.conf
    - --authorization-kubeconfig=/etc/kubernetes/scheduler.conf
    - --bind-address=127.0.0.1
    - --kubeconfig=/etc/kubernetes/scheduler.conf
    - --leader-elect=true
    image: registry.k8s.io/kube-scheduler:v1.30.0
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 8
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10259
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    name: kube-scheduler
    resources:
      requests:
        cpu: 100m
    startupProbe:
      failureThreshold: 24
      httpGet:
        host: 127.0.0.1
        path: /healthz
        port: 10259
        scheme: HTTPS
      initialDelaySeconds: 10
      periodSeconds: 10
      timeoutSeconds: 15
    volumeMounts:
    - mountPath: /etc/kubernetes/scheduler.conf
      name: kubeconfig
      readOnly: true
    - mountPath: /etc/kubernetes/allow-list.yaml
      name: allowlist
      readOnly: true
  hostNetwork: true
  priority: 2000001000
  priorityClassName: system-node-critical
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  volumes:
  - hostPath:
      path: /etc/kubernetes/scheduler.conf
      type: FileOrCreate
    name: kubeconfig
  - hostPath:
      path: /etc/kubernetes/allow-list.yaml
      type: File
    name: allowlist
```
Then check the pod's log
```
root@dev-control-plane:/etc/kubernetes/manifests# kubectl logs -n kube-system kube-scheduler-dev-control-plane -f
```

### Anything else we need to know?

It should be an easy fix by removing the locking logic https://github.com/kubernetes/kubernetes/blob/c15581b277e9e42d830898357fdbd92dcf7981c6/staging/src/k8s.io/component-base/metrics/opts.go#L366 and adding more test to cover this.

Also, I've sent a PR to update the documentation to apply the correct manifest format 
https://github.com/kubernetes/website/pull/48283

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
Client Version: v1.30.0
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by yongruilin**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
It removes unnecessary lock when parsing metrics allow list manifest.
It also adds a unit test to cover the related function.

#### Which issue(s) this PR fixes:

Fixes #127992 
[KEP]: https://github.com/kubernetes/enhancements/issues/2305

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]https://github.com/kubernetes/enhancements/issues/2305
```

### CODE REVIEW

File: staging/src/k8s.io/component-base/metrics/opts_test.go [hunk 1]
```diff
@@ -138,3 +140,68 @@ func TestConstrainLabelMap(t *testing.T) {
...
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			labelValueAllowLists = map[string]*MetricLabelAllowList{}
+			manifestFilePath := "/non-existent-file.yaml"
+			if tc.manifestExist {
+				tempFile, err := os.CreateTemp("", "allow-list-test")
+				if err != nil {
+					t.Fatalf("failed to create temp file: %v", err)
+				}
+				defer os.Remove(tempFile.Name())
```
Comment (rexagod) [inline staging/src/k8s.io/component-base/metrics/opts_test.go:193]: Needs error handling.

File: staging/src/k8s.io/component-base/metrics/opts_test.go [hunk 2]
```diff
@@ -138,3 +140,68 @@ func TestConstrainLabelMap(t *testing.T) {
...
+
+				if _, err := tempFile.WriteString(tc.manifest); err != nil {
+					t.Fatalf("failed to write to temp file: %v", err)
+				}
+				manifestFilePath = tempFile.Name()
+			}
+
+			SetLabelAllowListFromManifest(manifestFilePath)
+			if !cmp.Equal(labelValueAllowLists, tc.expectlabelValueAllowLists, cmp.AllowUnexported(MetricLabelAllowList{})) {
```
Comment (rexagod) [inline staging/src/k8s.io/component-base/metrics/opts_test.go:202]:
```suggestion
			if !reflect.DeepEqual(labelValueAllowLists, tc.expectlabelValueAllowLists) {
```

**Review by rexagod — CHANGES_REQUESTED** [review]
Thank you for the contribution!

File: staging/src/k8s.io/component-base/metrics/opts_test.go (hunk 1 above)
Comment (yongruilin) [reply staging/src/k8s.io/component-base/metrics/opts_test.go]: Updated. However, I see most of the case in the repo doesn't handle error when removing the temp file in test.

File: staging/src/k8s.io/component-base/metrics/opts_test.go (hunk 2 above)
Comment (yongruilin) [reply staging/src/k8s.io/component-base/metrics/opts_test.go:206]: I recommend we use `cmp.Equal` here as [cmp package](https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/google/go-cmp/cmp/compare.go#L7) is more powerful alternative to `reflect.DeepEqual`

Comment (rexagod) [reply staging/src/k8s.io/component-base/metrics/opts_test.go:206]: In the context of the operation being done here, I believe the two fulfill the same purpose, however, as a thumb rule, I'd prefer standard library over an external package, wherever possible.

Comment (yongruilin) [reply staging/src/k8s.io/component-base/metrics/opts_test.go]: I see. To be consistent with the existing test case,  I've updated to use `reflect.DeepEqual`.

**Comment by rexagod** [comment]
/approve
/lgtm

Thanks!

**Pull request merged**